File: src/main/java/io/github/chaosdave34/kitpvp/abilities/AbilityRunnable.java
package io.github.chaosdave34.kitpvp.abilities;

import io.github.chaosdave34.kitpvp.ExtendedPlayer;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;
import org.jetbrains.annotations.NotNull;

public abstract class AbilityRunnable extends BukkitRunnable {
    private final ExtendedPlayer extendedPlayer;
    private final ExtendedPlayer.GameType gameType;
    private final int deathCount;

    protected AbilityRunnable(ExtendedPlayer extendedPlayer) {
        this.extendedPlayer = extendedPlayer;
        this.gameType = extendedPlayer.getCurrentGame();
        this.deathCount = extendedPlayer.getTotalDeaths(gameType);
    }

    protected AbilityRunnable(Player p) {
        this(ExtendedPlayer.from(p));
    }

    @Override
    public void run() {
        if (!extendedPlayer.inGame()) cancel();
        if (gameType != extendedPlayer.getCurrentGame()) cancel();
        if (deathCount != extendedPlayer.getTotalDeaths(gameType)) cancel();
        runInGame();
    }

    public abstract void runInGame();

    public static BukkitTask runTaskLater(@NotNull Plugin plugin, @NotNull Runnable task, ExtendedPlayer extendedPlayer, long delay) {
        return new AbilityRunnable(extendedPlayer) {
            @Override
            public void runInGame() {
                task.run();
            }
        }.runTaskLater(plugin, delay);
    }

    public static BukkitTask runTaskLater(@NotNull Plugin plugin, @NotNull Runnable task, Player player, long delay) {
        return runTaskLater(plugin, task, ExtendedPlayer.from(player), delay);
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/abilities/AbilityRunnable.kt
package io.github.chaosdave34.kitpvp.abilities

import io.github.chaosdave34.kitpvp.ExtendedPlayer
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scheduler.BukkitTask
import java.util.*

abstract class AbilityRunnable(val uuid: UUID) : BukkitRunnable() {
    constructor(player: Player) : this(player.uniqueId)
    constructor(extendedPlayer: ExtendedPlayer) : this(extendedPlayer.uuid)

    private val gameType: ExtendedPlayer.GameType = ExtendedPlayer.from(uuid).currentGame
    private val deathCount: Int = ExtendedPlayer.from(uuid).getTotalDeaths(gameType)

    protected var i = 0

    override fun run() {
        val extendedPlayer = ExtendedPlayer.from(uuid)
        if (extendedPlayer.getPlayer() == null) return
        if (!extendedPlayer.inGame()) cancel()
        if (gameType != extendedPlayer.currentGame) cancel()
        if (deathCount != extendedPlayer.getTotalDeaths(gameType)) cancel()
        runInGame()
        i++
    }

    abstract fun runInGame()

    companion object {
        fun runTaskLater(plugin: Plugin, task: Runnable, extendedPlayer: ExtendedPlayer, delay: Long): BukkitTask {
            return object : AbilityRunnable(extendedPlayer) {
                override fun runInGame() {
                    task.run()
                }
            }.runTaskLater(plugin, delay)
        }

        fun runTaskLater(plugin: Plugin, task: Runnable, player: Player, delay: Long): BukkitTask {
            return runTaskLater(plugin, task, ExtendedPlayer.from(player), delay)
        }
    }
}